src/librustdoc/clean/mod.rs

```diff
@@ -250,13 +250,72 @@ pub(crate) fn clean_middle_region<'tcx>(region: ty::Region<'tcx>) -> Option<Life
     predicate: &hir::WherePredicate<'tcx>,
     cx: &mut DocContext<'tcx>,
+    parent_params: &[GenericParamDef],
+    parent_where_predicates: &[WherePredicate],
 ) -> Option<WherePredicate> {
     if !predicate.in_where_clause() {
         return None;
     }
+
     Some(match *predicate {
+        // There can be duplicates with parent predicates in this case so we dedup them. For
+        // example:
+        //
+        // ```
+        // pub trait Foo<T: Debug> {
+        //     fn bar() where T: Debug {} // This is valid but completely unneeded as `T` already
+        //                                // has this restriction.
+        //     fn foo() where T: Display {} // This item will only be available if `Self` implements
+        //                                  // both `Debug` and `Display`.
+        // }
+        // ```
         hir::WherePredicate::BoundPredicate(ref wbp) => {
-            let bound_params = wbp
+            let ty = clean_ty(wbp.bounded_ty, cx);
+
+            let (check_not_in_parent_bounds, check_not_in_parent_params): (
+                Box<dyn Fn(&GenericBound) -> bool>,
+                Box<dyn Fn(&Lifetime) -> bool>,
+            ) = if let Some((bounds, bound_params)) = parent_where_predicates
+                .iter()
+                .filter_map(|pred| match pred {
+                    WherePredicate::BoundPredicate { ty: parent_ty, bounds, bound_params }
+                        if *parent_ty == ty =>
+                    {
+                        Some((bounds, bound_params))
+                    }
+                    _ => None,
+                })
+                .next()
+            {
+                (
+                    Box::new(move |bound: &GenericBound| !bounds.contains(&bound)),
+                    Box::new(move |bound_param: &Lifetime| !bound_params.contains(&bound_param)),
+                )
+            } else {
+                (Box::new(|_: &GenericBound| true), Box::new(|_: &Lifetime| true))
+            };
+
+            let (check_not_in_parent_tys, check_not_in_parent_lifetimes): (
+                Box<dyn Fn(&GenericBound) -> bool>,
+                Box<dyn Fn(&Lifetime) -> bool>,
+            ) = if let Some(parent_bound) =
+                parent_params.iter().filter(|gen| Type::Generic(gen.name) == ty).next()
+            {
+                (
+                    Box::new(move |bound: &GenericBound| match &parent_bound.kind {
+                        GenericParamDefKind::Type { bounds, .. } => !bounds.contains(&bound),
+                        _ => true,
+                    }),
+                    Box::new(move |bound: &Lifetime| match &parent_bound.kind {
+                        GenericParamDefKind::Lifetime { outlives } => !outlives.contains(&bound),
+                        _ => true,
+                    }),
+                )
+            } else {
+                (Box::new(|_: &GenericBound| true), Box::new(|_: &Lifetime| true))
+            };
```

Comment:
This whole code isn't great but I don't know how to make it a bit better. If someone has an idea...

Reply:
What if instead of creating closures closing over `bounds` / `bound_params` / `outlives`, you simply return those aforementioned variables from the two `if let`s (returning `Default::default()` / `None` / `Vec::new()` from the `else` branch) and later write closures inline: `.filter(check_not_in_parent_bounds)` → `.filter(|bound| !bounds.contains(bound))`. Cloning should not be necessary since you should be able to use references. Not sure if that works out well and looks better but might be worth a shot.

Reply:
To illustrate:

```patch
- let (check_not_in_parent_bounds, check_not_in_parent_params): (
-     Box<dyn Fn(&GenericBound) -> bool>,
-     Box<dyn Fn(&Lifetime) -> bool>,
- ) = if let Some((bounds, bound_params)) =
-     parent_where_predicates.iter().find_map(|pred| match pred {
-         WherePredicate::BoundPredicate { ty: parent_ty, bounds, bound_params }
-             if *parent_ty == ty =>
-         {
-             Some((bounds, bound_params))
-         }
-         _ => None,
-     }) {
-     (
-         Box::new(move |bound: &GenericBound| !bounds.contains(&bound)),
-         Box::new(move |bound_param: &Lifetime| !bound_params.contains(&bound_param)),
-     )
- } else {
-     (Box::new(|_: &GenericBound| true), Box::new(|_: &Lifetime| true))
- };
+ let (parent_bounds, parent_bound_params) = parent_where_predicates
+     .iter()
+     .find_map(|pred| match pred {
+         WherePredicate::BoundPredicate {
+             ty: parent_ty,
+             bounds,
+             bound_params,
+         } if *parent_ty == ty => Some((bounds, bound_params)),
+         _ => None,
+     })
+     .unwrap_or_default();
```

Reply:
`bounds` & `bound_params` are of type `&Vec<_>` as far as I can tell, so I guess `.unwrap_or_default()` does not work (`Vec<_>` impls `Default` but `&Vec<_>` does not). However, you could instead write sth. like `.unwrap_or(|| const { (&Vec::new(), &Vec::new()) })` instead where the const block `const { … }` creates `&'static Vec<_>`s. Instead of const blocks, you could also use const items or apart from that `Cow::Owned(Vec::new())` (and `Cow::Borrowed` in the `Some` case.